ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -24,90 +27,253 @@
+    if (ensureFullyRead && stream.read() != -1) {
+      throw new IllegalArgumentException("Stream should have no more available bytes to read");
+    }
   }
 
   /**
-   * Re-constructs existing blobId by deserializing from data input stream
+   * Re-constructs existing blobId by deserializing from BlobId "string".
    *
-   * @param stream from which to deserialize the blobid
-   * @param clusterMap of the cluster that the blob id belongs to
+   * @param id of Blob as output by BlobId.getID().
+   * @param clusterMap of the cluster that the blobId belongs to.
    * @throws IOException
    */
-  public BlobId(DataInputStream stream, ClusterMap clusterMap) throws IOException {
-    this(stream, clusterMap, false);
+  public static BlobId fromStringId(String id, ClusterMap clusterMap) throws IOException {
+    return new BlobId(new DataInputStream(new ByteBufferInputStream(ByteBuffer.wrap(Base64.decodeBase64(id)))),
+        clusterMap, true);
   }
 
   /**
-   * Re-constructs existing blobId by deserializing from data input stream. This constructor includes an optional check
-   * that the stream has no more available bytes after reading.
+   * Re-constructs existing blobId by deserializing from data input stream
    *
-   * @param stream from which to deserialize the blobid
-   * @param clusterMap of the cluster that the blob id belongs to
-   * @param ensureFullyRead {@code true} if the stream should have no more available bytes after deserializing the blob
-   *                        ID.
+   * @param stream from which to deserialize the blobId.
+   * @param clusterMap of the cluster that the blobId belongs to.
    * @throws IOException
    */
-  private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead) throws IOException {
-    this.version = stream.readShort();
-    if (version == 1) {
-      partitionId = clusterMap.getPartitionIdFromStream(stream);
-      if (partitionId == null) {
-        throw new IllegalArgumentException("Partition ID cannot be null");
-      }
-      uuid = Utils.readIntString(stream);
-      if (ensureFullyRead && stream.read() != -1) {
-        throw new IllegalArgumentException("Stream should have no more available bytes to read");
-      }
-    } else {
-      throw new IllegalArgumentException("version " + version + " not supported for blob id");
-    }
+  public static BlobId fromDataInputStream(DataInputStream stream, ClusterMap clusterMap) throws IOException {
+    return new BlobId(stream, clusterMap, false);
   }
 
+  /**
+   * Size of blobId when it is serialized into bytes.
+   * @return The byte count of the serialized blobId.
+   */
   public short sizeInBytes() {
-    return (short) (Version_Size_In_Bytes + partitionId.getBytes().length + UUID_Size_In_Bytes + uuid.length());
+    switch (version) {
+      case BLOB_ID_V1:
+        return (short) (VERSION_FIELD_LENGTH_IN_BYTES + partitionId.getBytes().length + UUID_SIZE_FIELD_LENGTH_IN_BYTES
```

Comment:
Thought a bit and would rather keep the current way. Could not save many code nor increase readability.